wai-handler-launch/Network/Wai/Handler/Launch.hs

```diff
@@ -192,20 +193,41 @@ runUrlPort = runHostPortUrl "*4"
 
 runHostPortUrl :: String -> Int -> String -> Application -> IO ()
 runHostPortUrl host port url app = do
-    x <- newIORef True
-    _ <- forkIO $ Warp.runSettings
-        ( Warp.setPort port
-        $ Warp.setOnException (\_ _ -> return ())
-        $ Warp.setHost (fromString host) Warp.defaultSettings)
-        $ ping x app
-    launch port url
-    loop x
+    ready <- newIORef False
+    active <- newIORef True
+    let settings =
+          Warp.setPort port $
+          Warp.setOnException (\_ _ -> return ()) $
+          Warp.setHost (fromString host) $
+          Warp.setBeforeMainLoop (writeIORef ready True) $
+          Warp.defaultSettings
+    withAsync
+      -- server thread, will be killed if we terminate
+      (Warp.runSettings settings (ping active app))
+      (\a -> do
+          -- launch & monitor thread, reraises exceptions from server thread
+          link a
+
+          -- launch the browser, after waiting until the server
+          -- signals that it's ready (or raises a startup exception)
+          let wait = do
```

Comment:
Why use an IORef here? If you use an MVar, then you can get blocking behavior and not need to worry about polling.